src/section/mod.rs

```diff
@@ -8,27 +8,395 @@
+        self.members
+            .all()
+            .filter(move |info| {
+                self.members.is_joined(info.peer.name()) || self.is_elder(info.peer.name())
+            })
+            .map(|info| &info.peer)
+    }
+
+    /// Returns adults from our section.
+    pub fn adults(&self) -> impl Iterator<Item = &Peer> {
+        self.members
+            .adults()
+            .filter(move |peer| !self.is_elder(peer.name()))
+    }
+
+    pub fn find_member_from_addr(&self, addr: &SocketAddr) -> Option<&Peer> {
+        self.members
+            .all()
+            .find(|info| info.peer.addr() == addr)
+            .map(|info| &info.peer)
+    }
+
+    // Returns age of a member with `name` or `MIN_AGE` if not found.
+    pub fn member_age(&self, name: &XorName) -> u8 {
+        self.members
+            .get(name)
+            .map(|info| info.peer.age())
+            .unwrap_or(MIN_AGE)
+    }
+
+    fn elders_info_signing_key_index(&self) -> u64 {
+        // NOTE: we assume that the key the current `EldersInfo` is signed with is always
+        // present in our section proof chain. This is currently guaranteed, because we use the
+        // `SectionUpdateBarrier` and so we always update the current `EldersInfo` and the current
+        // section key at the same time.
+        self.chain
+            .index_of(&self.elders_info.proof.public_key)
+            .unwrap_or_else(|| unreachable!("EldersInfo signed with unknown key"))
+    }
+
+    // Tries to split our section.
+    // If we have enough mature nodes for both subsections, returns the elders infos of the two
+    // subsections. Otherwise returns `None`.
+    fn try_split(
+        &self,
+        network_params: &NetworkParams,
+        our_name: &XorName,
+    ) -> Option<(EldersInfo, EldersInfo)> {
+        let next_bit_index = if let Ok(index) = self.prefix().bit_count().try_into() {
+            index
+        } else {
+            // Already at the longest prefix, can't split further.
+            return None;
+        };
+
+        let next_bit = our_name.bit(next_bit_index);
+
+        let (our_new_size, sibling_new_size) = self
+            .members
+            .adults()
+            .map(|peer| peer.name().bit(next_bit_index) == next_bit)
+            .fold((0, 0), |(ours, siblings), is_our_prefix| {
+                if is_our_prefix {
+                    (ours + 1, siblings)
+                } else {
+                    (ours, siblings + 1)
+                }
+            });
+
+        // If either of the two new sections will not contain enough entries, return `false`.
+        if our_new_size < network_params.recommended_section_size
+            || sibling_new_size < network_params.recommended_section_size
+        {
+            return None;
+        }
+
+        let our_prefix = self.prefix().pushed(next_bit);
+        let other_prefix = self.prefix().pushed(!next_bit);
+
+        let our_elders = self.members.elder_candidates_matching_prefix(
```

Comment:
looks like here is only place calling `elder_candidates_matching_prefix` twice with different prefix.
why not have one function, say `split_candidates`, which returns pair result directly?

Reply:
We could do that, but I don't think it would make much difference. The logic would still exist, just in a different place (in `SectionPeers`).